### ISSUE TITLE
[ENH] Please stop using undocumented CPython internals.

### ISSUE THREAD

**Issue by markshannon** [cython/cython#4382]
The use of undocumented CPython internals in Cython causes subtle bugs and makes in unnecessarily difficult for Cython based modules to be used across different versions of (C)Python.

The latest example of a breakage (amongst many) is https://bugs.python.org/issue43760

We[1] are happy to consider API extensions to support Cython. We regard Cython as an important part of the Python ecosystem, and would love to cooperate, rather than conflict with Cython.
This cooperation has proved fruitful in the past.
For example, [PEP 590](https://www.python.org/dev/peps/pep-0590/) was designed to suuport Cython generated callables. Because it had to be designed to be both fast and general it has proved beneficial to CPython performance as well.

Cython has many  `__Pyx_Blah` functions which access CPython internals. Moving these to the CPython C-API has a lot of advantages for both CPython and Cython:

* Improved reliability of Cython generated code. Subtle changes in the implicit semantics of CPython internals will not break Cython code in non-obvious ways.
* Improved portability of Cython modules across Python versions.
* Reduced maintenance for Cython: Just use the CPython API; no need to worry about CPython internals
* Reduced maintenance for CPython: We can treat C structs as opaque and not worry about breakage, as long as we maintain the C API functions

#### Performance

I assume that there are two reasons that Cython uses the direct access to CPython internals:
* There is no suitable API. As stated above, if there is no suitable API for a given purpose then please propose one[2].
* For performance. 

Regarding performance, a well designed API may well be faster than probing and modifying CPython internals. It allows the CPython VM to make decisions as when to modify state lazily and when to cache values.

There will be a few things where the overhead of a call would be too high, but they should be very few, and understood by both CPython developers and Cython developers. The obvious example is `Py_INCREF` and `Py_DECREF`.

[1] The CPython developers. I hope the other CPython developers don't mind me speaking for them.

[2] Please propose APIs at as high and as abstract a level as possible. E.g don't propose an API for inserting Python frames in the frame stack, and for creating code objects. Propose an API for inserting a frame into the frame stack with writable location information.

**Comment by scoder** [comment]
Thank you for your comments, Mark. Let me clarify a couple of points.

There are certainly two types of Cython users. Those who only want to generate their C code and have it work across CPython versions, and those who additionally care about performance. For the first, your requests seem reasonable. For the latter, not so much.

> Propose an API for inserting a frame into the frame stack with writable location information.

That's pretty much what happened in [this recent discussion](https://mail.python.org/archives/list/python-dev@python.org/thread/ZWTBR5ESYR26BUIVMXOKPFRLGGYDJSFC/) on the python-dev mailing list.

> Cython has many `__Pyx_Blah` functions which access CPython internals. Moving these to the CPython C-API has a lot of advantages for both CPython and Cython

A few of these were written to mimic CPython features that do not have a (suitable) C-API, but many *are* part of the C-API, in one way or another. Most of them exist for performance reasons and in order to be inlined in Cython. They are unlikely to go away.

However, when we touch internals, we usually make it optional by providing alternative implementations – if possible. As long as there is some kind of official C-API way to do things, we'll have at least a fallback implementation that does them that way, if only to cater for PyPy and friends. The generated C code automatically selects the right implementation based on the Python implementation and version. If a newer CPython version has specific needs or restrictions, we can accommodate it in that infrastructure. See the list of C compile time options [here](https://github.com/cython/cython/blob/e2a23fe11063ee13c886e89007207a07a397e016/Cython/Utility/ModuleSetupCode.c#L56). I think this is an excellent way to cater for all sides.

Now, previously, it used to be the case that CPython's C-API was more stable than Cython. This seems to have shifted. Over the years, I was advocating for generating the C code from Cython modules locally and shipping it as part of the source distribution. As long as the C-API (or the exposed internals) were stable enough, that was a good idea. If they no longer are, then it's less of a good idea. The C code can get outdated and it's then really hard for users to regenerate it when all they want to do is pip-install a package into a new CPython version for which there's no wheel (and possibly never will be, for an unmaintained package). This was a problem with the `line_profiler` package, for example, which, for a very long while, didn't receive an update with freshly generated C sources to support CPython 3.7, and the C sources failed to compile on that newly released CPython version at the time.

I'm wondering if setuptools can help with this. If it gained a "forced recythonize" option to override existing C sources and recreate them from scratch, using a freshly installed Cython, then users would gain a much simpler upgrade path.

Also, wheels work very smoothly these days, so maybe it's generally time to accept a complete regenerate-rebuild cycle for all platforms for which there is no distribution wheel?

I really think that the underlying issues can largely be solved on the distribution side, rather than letting Cython generate less adaptive code.

**Comment by markshannon** [comment]
I have no doubt that parts of the C-API are too slow for Cython, I mentioned `Py_DECREF` and `Py_INCREF` early.
However, you seem to be taking a dogmatic view that all calls to the C-API are too slow and must be replaced with Cython versions.

What evidence do you have for this?
Which parts of the C-API are too slow?

If you want any new APIs please file an issue. Suggestions in the middle of a email thread just get lost, I'm afraid.

**Comment by scoder** [comment]
> you seem to be taking a dogmatic view that all calls to the C-API are too slow and must be replaced with Cython versions.

No, certainly not. But I am questioning that exposing and using internals between CPython and Cython is a bad thing per se. I personally think that it's great to be able to interface at a low level and to hook directly into existing data structures. After all, there _is_ an implementation and not just an interface. That implementation might change, but so might the interfaces to it. Both have happened before, and both are happening right now.

Cython is a code generator, and as such, I think it can afford to take advantage of deeper implementation details – not only of CPython version x.y, but of whatever Python implementation in whatever version it compiles on.

I suggest to think of Cython as a replacement of the C-API, rather than a user of the C-API, and of the code that it generates as an integration into CPython/PyPy/GraalVM/… rather than an external plugin.

As I wrote before:
> I really think that the underlying issues can largely be solved on the distribution side, rather than letting Cython generate less adaptive code.

**Comment by robertwb** [comment]
I'll echo that we do try to use the public C-API and only rely on internals/our own implementations when they are insufficiently flexible and/or there are are material, measured performance gains (for some still-supported versions of Python at least). The fact that this is generated code not intended for human consumption, and that its widespread use has a multiplicative effect for every optimization, means the balance is different than for hand-written code. 

I, too, would like to see better support for regenerating it on the distribution side rather than a maintainer having to ship the .c files pinned to some specific Cython version and manual efforts to get packages to "recythonize" for new Python releases.

**Comment by dgrunwald** [comment]
I think it's fine if Cython uses implementations details for performance, but it should do so only in `#if PY_VERSION_HEX < 0x030B0000` blocks. That way the generated code falls back on the documented API if it's used with a future Python version that that Cython version wasn't tested against.

This would allow CPython to freely change the implementation details without waiting for Cython; and users would lose performance only temporarily until Cython has caught up.

**Comment by scoder** [comment]
FWIW, I was recently asked what the performance difference is between the Limited API and CPython-optimised Cython, and had no answer. I then tried it with a simple caching example that uses ints, dicts and lists, and got a bit more than a 10% slowdown for the Limited API case. So, there are visible differences in the tricks that Cython pulls in the C-API, but in the end, these tricks have to add up to reach that difference, and are obviously highly code specific.

**Comment by rgommers** [comment]
> I, too, would like to see better support for regenerating it on the distribution side rather than a maintainer having to ship the .c files pinned to some specific Cython version and manual efforts to get packages to "recythonize" for new Python releases.

This is a separate issue from the main topic here I'd say, but: please stop recommending this, it is definitely bad practice in 2022-23. It used to be necessary 4+ years ago (and we did it in NumPy, SciPy, et al.), but now that we have `pyproject.toml` with reliable a build dependency section to put Cython into, there is no reason anymore to ship C code and it has become an anti-pattern (because outdated C code, as pointed out higher up). 

For more distribution-related context: the historical reason to ship generated C sources was that we only had `build_requires=` in `setuptools`, and using that would invoke `easy_install`, leading to all sorts of hairy issues. `pip` did not support build dependencies back then. Right now it does, and `easy_install` is gone. So all you need is a `pyproject.toml` file contaning:
```toml
[build-system]
requires = ["Cython"]  # or with version control: `"Cython>=0.29.32,<3.0"
```
and you're all set.

tl;dr the distribution issues you are pointing at have been solved.

**Comment by leofang** [comment]
@rgommers Yes, the plan is to revert the recommendation, see #5089.
